### ISSUE TITLE
Backup Restore service goes into quorum loss when backup policy is created with retention duration more than 24 days

### ISSUE THREAD

**Issue by khandelwalbrijeshiitr** [Microsoft/service-fabric#269]
On SF runtime version v6.4 or above, there is an issue when you create a backup policy with retention duration more than 24 days. The create policy API fails immediatly however due to some bug in the cleanup, the information stays somewhere in store. As a result now, once the FabricBRS primary fails over, it would fail to come up as it tries to initialize the System.Timers.Timer with internal value which is more than Int32.MaxValue and keeps failing. This results in service going into quorum loss. 

The error you can see in explorer is: 

"RunAsync failed due to an unhandled exception causing the host   process to crash: System.ArgumentOutOfRangeException: Number must be either   non-negative and less than or equal to Int32.MaxValue or -1.            Parameter name:   dueTime               at   System.Threading.Timer..ctor(TimerCallback callback, Object state, Int32   dueTime, Int32 period)               at   System.Timers.Timer.set_Enabled(Boolean value)               at   System.Fabric.BackupRestore.BRSTimer.ArmTimer(Double intervalInMilliSeconds)               at   System.Fabric.BackupRestore.Service.RetentionTimer.armTimer(Boolean   postFailedReschedule, RetentionMetadata retentionMetadata, Boolean   updatePolicyToAddRetention)               at   System.Fabric.BackupRestore.Service.RetentionManager.d__18.MoveNext()            --- End of   stack trace from previous location where exception was thrown ---               at   System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()               at   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task   task)               at   System.Fabric.BackupRestore.Service.BackupRestoreService.d__10.MoveNext()            --- End of   stack trace from previous location where exception was thrown ---               at   System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()               at   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task   task)               at   Microsoft.ServiceFabric.Services.Runtime.StatefulServiceReplicaAdapter.d__23.MoveNext()"

It is a bug in version 6.4 of service fabric.

**Comment by khandelwalbrijeshiitr** [comment]
To mitigate, we need to clear the incorrect data from Fabric BRS. The following steps would incur data loss on Fabric BRS which would mean existing policies configured in FabricBRS would be lost. Before applying this mitigation this should be clarified with the customer and only then one should process. Not any backups in the storage would stay intact and once the policies are configured again, one would be able to enumerate them. 
Steps: 
1. Check if the FabricBRS is in quorum loss state.
2. Update the service to have MinReplicaSetSize and TargetReplicaSetSize to 1 and also update ReplicaRestartWaitDuration and QuorumLossWaitDuration to 30.
This can be done using below command.

```powershell
    Update-ServiceFabricService -Stateful -ServiceName "fabric:/System/BackupRestoreService" -TargetReplicaSetSize 1 -MinReplicaSetSize 1  -ReplicaRestartWaitDuration 30 -QuorumLossWaitDuration 30
```

3. Wait for 3 -4 mins till the quorum loss error goes away on the cluster and the primary of FabricBRS is healthy.
4. After this step, update the service using the same command to default values (assuming the service was running with 5 replicas earlier, if not, update the replica set size values accordingly as per other system services).

 ``` powershell 
    Update-ServiceFabricService -Stateful -ServiceName "fabric:/System/BackupRestoreService" -TargetReplicaSetSize 5 -MinReplicaSetSize 3 -ReplicaRestartWaitDuration 1800 -QuorumLossWaitDuration 172800
```
5. After this step FabricBRS should run with 5 replicas and should be healthy. 
6. Customer should be advised that he\she can create the backup policy with same storage and schedule but with retentionPolicy with retentionDuration less than 24 days.In this case, no backups are lost. Customer can enumerate and restore from all the backups created.

**Comment by raunakpandya** [comment]
The fix is now released in the latest Service Fabric v6.4 refresh. https://blogs.msdn.microsoft.com/azureservicefabric/2019/02/04/azure-service-fabric-6-4-refresh-release/
